env/env_test.cc

```diff
@@ -839,6 +839,64 @@ TEST_P(EnvPosixTestWithParam, DecreaseNumBgThreads) {
   WaitThreadPoolsEmpty();
 }
 
+TEST_P(EnvPosixTestWithParam, ReserveThreads) {
+  constexpr int kWaitMicros = 10000000;  // 10seconds
+  std::vector<test::SleepingBackgroundTask> tasks(4);
+  // Set number of thread to 4 first.
+  env_->SetBackgroundThreads(4, Env::Priority::HIGH);
+  // Wait until all threads start (num_waiting accumulates)
+  // Reserve 3 threads
+  Env::Default()->SleepForMicroseconds(kDelayMicros);
+  ASSERT_EQ(3, env_->ReserveThreads(3, Env::Priority::HIGH));
+
+  // Schedule 3 tasks. 0 running; Task 1, 2 waiting; 3 reserved threads.
+  for (size_t i = 0; i < 3; i++) {
+    env_->Schedule(&test::SleepingBackgroundTask::DoSleepTask, &tasks[i],
+                   Env::Priority::HIGH);
+  }
+  ASSERT_FALSE(tasks[0].TimedWaitUntilSleeping(kWaitMicros));
+  ASSERT_EQ(2U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+  ASSERT_TRUE(tasks[0].IsSleeping());
+  ASSERT_TRUE(!tasks[1].IsSleeping());
+  ASSERT_TRUE(!tasks[2].IsSleeping());
+
+  // Release 2 threads. Task 0, 1, 2 running; 1 reserved thread.
+  ASSERT_EQ(2, env_->ReleaseThreads(2, Env::Priority::HIGH));
+  ASSERT_FALSE(tasks[1].TimedWaitUntilSleeping(kWaitMicros));
+  ASSERT_FALSE(tasks[2].TimedWaitUntilSleeping(kWaitMicros));
+  ASSERT_EQ(0U, env_->GetThreadPoolQueueLen(Env::Priority::HIGH));
+  ASSERT_TRUE(tasks[1].IsSleeping());
+  ASSERT_TRUE(tasks[2].IsSleeping());
+
+  // Task 1, 2 running; 1 reserved thread.
+  tasks[0].WakeUp();
+  ASSERT_FALSE(tasks[0].TimedWaitUntilDone(kWaitMicros));
+  Env::Default()->SleepForMicroseconds(kDelayMicros);
```

Comment:
Should we add `ASSERT_EQ(0, env_->ReserveThreads(3, Env::Priority::HIGH));` as a prior before any change that can lead to  `ASSERT_EQ(1, env_->ReserveThreads(3, Env::Priority::HIGH));` below to prove the point of `1 more thread can be reserved`?

Reply:
We have it at line 850.

Reply:
No problem, I will add it.

Reply:
Yes, will add it.